llvm/lib/Target/AIE/ReservedRegsLICM.cpp

```diff
@@ -0,0 +1,330 @@
+  bool tryHoistToPreHeader(const CandidateInfo &Cand, MachineLoop &L);
+
+  /// Hoist \p Cand out of the loop to \p InsertBefore.
+  void hoist(const CandidateInfo &Cand,
+             MachineBasicBlock::iterator InsertBefore,
+             MachineBasicBlock &InsertMBB);
+
+  /// Verify if \p Cand is loop invariant and can be safely hoisted.
+  /// \pre Cand->DefinedReg has a unique live value within the loop. This is
+  /// verified by processForExitHoist().
+  bool isLoopInvariantInst(const CandidateInfo &Cand, const MachineLoop &L);
+};
+
+} // end anonymous namespace
+
+char ReservedRegsLICM::ID;
+
+char &llvm::ReservedRegsLICMID = ReservedRegsLICM::ID;
+
+INITIALIZE_PASS_BEGIN(ReservedRegsLICM, DEBUG_TYPE,
+                      "Machine LICM for reserved regs", false, false)
+INITIALIZE_PASS_DEPENDENCY(MachineLoopInfo)
+INITIALIZE_PASS_END(ReservedRegsLICM, DEBUG_TYPE,
+                    "Machine LICM for reserved regs", false, false)
+
+llvm::FunctionPass *llvm::createReservedRegsLICMPass() {
+  return new ReservedRegsLICM();
+}
+
+bool ReservedRegsLICM::runOnMachineFunction(MachineFunction &MF) {
+  if (skipFunction(MF.getFunction()))
+    return false;
+
+  const TargetSubtargetInfo &ST = MF.getSubtarget();
+  TRI = ST.getRegisterInfo();
+  MRI = &MF.getRegInfo();
+  assert(MRI->isSSA() && "PhysRegCopyTracker can only track SSA copies");
+
+  LLVM_DEBUG(dbgs() << "******** Reserved register LICM: " << MF.getName()
+                    << " ********\n");
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  for (MachineLoop *L : MLI.getBase()) {
+    runOnLoop(*L);
+  }
+
+  return Changed;
+}
+
+static MCRegister getSinglePhysRegDef(const MachineInstr &MI) {
+  if (MI.getNumDefs() == 1 && MI.getNumExplicitDefs() == 1) {
+    const MachineOperand &MO = MI.getOperand(0);
+    return MO.isReg() && MO.getReg().isPhysical() ? MO.getReg().asMCReg()
+                                                  : MCRegister();
+  }
+  return MCRegister();
+}
+
+ReservedRegsLICM::CandidateInfo *
+ReservedRegsLICM::getCandidateInfo(const MachineInstr &MI) {
+  if (MCRegister Reg = getSinglePhysRegDef(MI)) {
+    auto It = Candidates.find(Reg);
+    if (It != Candidates.end()) {
+      return &It->second;
+    }
+    if (MRI->canSimplifyPhysReg(Reg)) {
+      // Only consider "simplifiable" reserved regs in this pass.
+      return &Candidates.try_emplace(Reg, Reg).first->second;
+    }
+  }
+  return nullptr;
+}
+
+BitVector ReservedRegsLICM::collectLoopReservedLiveins(const MachineLoop &L) {
+  const MachineBasicBlock &Header = *L.getHeader();
+  LivePhysRegs LiveRegs(*TRI);
+
+  // Conservativaly assume reserved reserved regs are all liveouts
+  for (MCPhysReg PhysReg : MRI->getReservedRegs().set_bits()) {
+    if (MRI->canSimplifyPhysReg(PhysReg)) {
+      LiveRegs.addReg(PhysReg);
+    }
+  }
+
+  // Traverse instructions to remove defs
+  for (const MachineInstr &MI : reverse(Header))
```

Comment:
I was thinking about one case that one register is defined by a block that dominates the Header. I think this case is not captured here, this register will not be in the live in set. 

Can we just trust `LoopBody->liveins()`? set